forte/processors/data_augment/algorithms/UDA.py

```diff
@@ -0,0 +1,192 @@
+            - :attr:`'mean'`: the output will be divided by
+                the number of elements in the output.
+    """
+    def __init__(
+            self,
+            sup_iterator: DataIterator,
+            unsup_iterator: DataIterator,
+            unsup_forward_fn: Callable,
+            softmax_temperature: float = 1.0,
+            confidence_threshold: float = -1,
+            reduction: str = "mean",
+    ):
+        self.sup_iterator: DataIterator = sup_iterator
+        self.unsup_iterator: DataIterator = unsup_iterator
+        self.unsup_forward_fn = unsup_forward_fn
+        self.softmax_temperature = softmax_temperature
+        self.confidence_threshold = confidence_threshold
+        self.reduction = reduction
+
+        # The flag for returning the unsupervised data.
+        self.use_unsup = True
+        self.sup_iter: Iterator[Batch]
+        self.unsup_iter: Iterator[Batch]
+
+    def __len__(self):
+        return self.sup_iterator.__len__()
+
+    def switch_to_dataset(
+            self,
+            dataset_name: Optional[str] = None,
+            use_unsup: bool = True
+    ):
+        # Set the flag of using unsupervised data.
+        self.use_unsup = use_unsup
+        self.sup_iterator.switch_to_dataset(dataset_name)
+
+    def switch_to_dataset_unsup(
+            self,
+            dataset_name: Optional[str] = None
+    ):
+        self.unsup_iterator.switch_to_dataset(dataset_name)
+
+    @property
+    def num_datasets(self) -> int:
+        return self.sup_iterator.num_datasets \
+               + self.unsup_iterator.num_datasets
+
+    @property
+    def dataset_names(self) -> List[str]:
+        return self.sup_iterator.dataset_names \
+               + self.unsup_iterator.dataset_names
+
+    def calculate_uda_loss(
+            self,
+            unsup_batch: Batch,
+            unsup_forward_fn: Callable
+    ) -> Tensor:
+        r"""
+        This function calculate the KL divergence
+        between the output probabilities of original
+        input and augmented input. It calls the user-provided
+        :func: unsup_forward_fn to perform the prediction.
+
+        Args:
+            - unsup_batch: A batch containing the unsupervised data.
+                There should be both original & augmented inputs in
+                each training example.
+            - unsup_forward_fn: The user-provided forward function
+                of the model.
+        Returns:
+            - The loss, as a pytorch scalar float tensor.
+        """
+        logits_orig, logits_aug = unsup_forward_fn(unsup_batch)
+        uda_loss = kl_divg_loss_with_logits(
+            target_logits=logits_orig,
+            input_logits=logits_aug,
+            softmax_temperature=self.softmax_temperature,
+            confidence_threshold=self.confidence_threshold,
+            reduction=self.reduction
+        )
+        return uda_loss
+
+    def __iter__(self):
+        self.sup_iter = iter(self.sup_iterator)
```

Comment:
Why are we setting up the two iterator here?

Reply:
The DataIterator in texar-pytorch is not inherited from the Iterator. To get the iterator, we have to explicitly call the __iter__ function. Check [this](https://github.com/asyml/texar-pytorch/blob/master/texar/torch/data/data/data_iterators.py#L546) for details.